server/src-lib/Hasura/GraphQL/Parser/Collect.hs

```diff
@@ -0,0 +1,283 @@
+  mergeFields =<< flattenSelectionSet objectTypeNames selectionSet
+
+-- | Flattens inline fragments in a selection set. For example,
+--
+-- > {
+-- >   bar
+-- >   ... on Foo {
+-- >     baz
+-- >     qux
+-- >   }
+-- > }
+--
+-- is flattened to:
+--
+-- > {
+-- >   bar
+-- >   baz
+-- >   qux
+-- > }
+--
+-- Nested fragments are similarly flattened, but only as is necessary: fragments
+-- inside subselection sets of individual fields are /not/ flattened. For
+-- example,
+--
+-- > {
+-- >   bar
+-- >   ... on Foo {
+-- >     baz {
+-- >       ... on Baz {
+-- >         foo
+-- >       }
+-- >     }
+-- >     qux
+-- >   }
+-- > }
+--
+-- is flattened to
+--
+-- > {
+-- >   bar
+-- >   baz {
+-- >     ... on Baz {
+-- >       foo
+-- >     }
+-- >   }
+-- >   qux
+-- > }
+--
+-- leaving the innermost fragment on @baz@ alone.
+--
+-- This function also applies @\@include@ and @\@skip@ directives, since they
+-- should be applied before fragments are flattened.
+flattenSelectionSet
+  :: (MonadParse m, Foldable t)
+  => t Name
+  -- ^ The name of the object type the 'SelectionSet' is selecting against.
+  -> SelectionSet NoFragments Variable
+  -> m [Field NoFragments Variable]
+flattenSelectionSet objectTypeNames = fmap concat . traverse flattenSelection
+  where
+    -- The easy case: just a single field.
+    flattenSelection (SelectionField field) = do
+      void $ validateDirectives (_fDirectives field)
+      applyInclusionDirectives (_fDirectives field) $ pure [field]
+    -- The 'SelectionFragmentSpread' case has already been eliminated by the fragment inliner.
+    -- The involved case: we have an inline fragment to process.
+    flattenSelection (SelectionInlineFragment fragment) = do
+      validateDirectives (_ifDirectives fragment)
+      let fields =
+            case _ifTypeCondition fragment of
+             -- No type condition, so the fragment unconditionally applies.
+              Nothing -> flattenInlineFragment fragment
+              Just typeName
+                -- There is a type condition, but it is just the type of the
+                -- selection set; the fragment trivially applies.
+                | typeName `elem` objectTypeNames -> flattenInlineFragment fragment
+
+                -- TODO(PDV) implement this check:
+                -- http://spec.graphql.org/June2018/#sec-Fragment-Spread-Type-Existence
+
+                -- TODO(PDV) we are now starting to support interfaces and unions; fix the below
+
+                -- Otherwise, the fragment must not apply, because we do not currently
+                -- support interfaces or unions. According to the GraphQL spec, it is
+                -- an *error* to select a fragment that cannot possibly apply to the
+                -- given type; see
+                -- http://spec.graphql.org/June2018/#sec-Fragment-spread-is-possible.
+                -- Therefore, we raise an error.
+                | otherwise -> return []
```

Comment:
Why does this not throw an error?

```graphql
fragment f on actors {
  first_name
}

query {
  tracks_connection {
    edges {
      node {
        ...f
      }
    }
  }
}
```
The above query should throw an error but goes through currently.

Reply:
It threw an error when I wrote it. 🙂 It looks like the error is now commented out, perhaps because we now *do* support interfaces and unions. I don’t fully understand the way they are implemented, but maybe @abooij can weigh in on what should be done here.